APE22.rst

```diff
@@ -0,0 +1,440 @@
+
+
+Let's consider an example::
+
+   src/
+      __init__.py::
+         __all__ = ["foo", "spam"]
+
+      foo.py::
+         __all__ = ["func"]
+         def func():
+            pass
+
+         def _func():
+            pass
+      
+      _bar.py::
+         __all__ = ['bunc']
+         def bunc():
+            pass
+
+         def _bunc():
+            pass
+
+      baz.py::
+         __all__ = ['qux']
+
+         def qux():
+            pass
+
+      spam/  # implicit namespace package
+
+         ham.py::
+            __all__ = ['eggs']
+
+            def eggs():
+               pass
+
+In this case, ``bar`` is public in ``foo``, and  ``foo`` is public in ``src``,
+so ``src.foo.func`` is public. This is public. In contrast, while ``bunc`` is
+public in ``_bar``, ``src._bar`` is not public, so ``src._bar.bunc`` is not
+public either. This is "locally" public, as in internal. Both ``_func`` and
+``_bunc`` are internal.
+
+=======  =========  ==========  ==========  =========================
+Symbol                                      Status                            
+=======  =========  ==========  ==========  =========================
+``src``                                     public
+         ``.foo``                           public
+                    ``.func``               public
+                    ``._func``              internal
+         ``._bar``                          public
+                    ``.bunc``               internal (locally public)
+                    ``._bunc``              internal
+         ``.baz``                           internal
+                    ``.bunc``               internal (locally public)
+                    ``._bunc``              internal
+         ``spam``                           public
+                    ``.ham``                public
+                                ``.eggs``   public
+=======  =========  ==========  ==========  =========================
+
+
+Branches and pull requests
+--------------------------
+
+N/A
+
+
+Implementation
+--------------
+
+The process of adopting this APE will change Astropy's API. This APE formally
+establishes Astropy's API, so what is meant by "change"?  There is already a *de
+facto* API, which is what users *think* is the API.  This is determined
+primarily by the documentation, but also by tab-completion within an interactive
+session.  In the implementation of this APE to establish a *formal* API, we aim
+to minimize changes to the *de facto* API. This will be accomplished in the
+following 5 phases:
+
+1. **Snapshot the documentation**. As of the adoption of the APE a snapshot of
+   the documentation will be saved. This will be used to determine what is
+   currently public and what is not. Until the APE's adoption is complete, this
+   snapshot is authoritative, e.g. dictating what must be added and removed from
+   ``__all__``, for deciding what must undergo a deprecation process, etc.
```

Comment:
I feel like we lack a prescription for what to do with symbols that are added post-snapshot here

Reply:
They should follow the rules! If it's public, make it so.
This is just saying we're not going to debate what already exists until we rectify the current ad-hoc standard with this one. But new things should follow the new standard.

Reply:
Agreed, but I think I meant to suggest that it should be explicitly mentioned in the APE too.